src/OpenFOAM/device/DeviceConfig.H

```diff
@@ -61,56 +61,58 @@ namespace Foam
 }
 
 
-
-template<>
-inline bool Foam::hasAtomicAdd<int>()
+namespace Foam
 {
-    return true;
-}
+    template<>
+    inline bool hasAtomicAdd<int>()
+    {
+        return true;
+    }
 
-template<>
-inline bool Foam::hasAtomicAdd<long>()
-{
-    return true;
-}
+    template<>
+    inline bool hasAtomicAdd<long>()
+    {
+        return true;
+    }
 
-template<>
-inline bool Foam::hasAtomicAdd<float>()
-{
-    return true;
-}
+    template<>
+    inline bool hasAtomicAdd<float>()
+    {
+        return true;
+    }
 
-template<>
-inline bool Foam::hasAtomicAdd<double>()
-{
-    return currentComputeCapability() >= 60;
-}
+    template<>
+    inline bool hasAtomicAdd<double>()
+    {
+        return currentComputeCapability() >= 60;
+    }
 
 
-template<>
-struct Foam::is_number<int>: std::true_type {};
+    template<>
+    struct Foam::is_number<int>: std::true_type {};
 
 
-template<>
-struct Foam::is_number<long>: std::true_type {};
+    template<>
+    struct Foam::is_number<long>: std::true_type {};
 
 
-template<>
-struct Foam::is_number<float>: std::true_type {};
+    template<>
+    struct Foam::is_number<float>: std::true_type {};
 
 
-template<>
-struct Foam::is_number<double>: std::true_type {};
+    template<>
+    struct Foam::is_number<double>: std::true_type {};
 
 
-template<class T>
-inline bool Foam::hasAtomicAdd()
-{
-    static_assert(is_number<T>::value, "Number is required");
-    return false;
-}
+    template<class T>
+    inline bool hasAtomicAdd()
+    {
+        static_assert(is_number<T>::value, "Number is required");
+        return false;
+    }
 
-#else
-#error "Currently only CUDA is supported."
-#endif
+    #else
+    #error "Currently only CUDA is supported."
+    #endif
 
+}
```

Comment:
The namespace closing bracket should be before the preprocessor directives.
Currently the namespace and preprocessor scopes itersect.